cli/listener/src/main.rs

```diff
@@ -570,18 +611,69 @@ fn get_success_response(interval_prices: IntervalPrices) -> ResponseBox {
     .boxed()
 }
 
+/// Get a header with `Expires` key.
+/// If the end date for the query is far in the past we set it to expire in 1
+/// week. However, if the end date is in the same epoch, we set it to expire in
+/// the approximate time the next epoch begins.
+fn get_expires_header(
+    now: chrono::DateTime<chrono::Utc>,
+    interval_prices: &IntervalPrices,
+    current_clock: Option<Clock>,
+    date_request_type: &DateRequestType,
+) -> Header {
+    let expires_date = match date_request_type {
+        // From the past, set expiration to 1 week from now.
+        DateRequestType::Past => now + chrono::Duration::weeks(1),
+        DateRequestType::MovingTarget => {
+            // We don't have information about the current clock, default to 1hr
+            // cache.
+            match current_clock {
+                None => now + chrono::Duration::hours(1),
+                Some(clock) => {
+                    let epoch_sched = EpochSchedule::without_warmup();
+                    let first_slot_next_epoch =
+                        epoch_sched.get_first_slot_in_epoch(clock.epoch + 1);
+                    let slots_diff = first_slot_next_epoch - clock.slot;
+                    if interval_prices.duration_epochs() > 0 {
+                        // We can estimate the epoch duration.
+                        let epoch_duration = interval_prices.duration_wall_time()
+                            / interval_prices.duration_epochs() as i32;
+                        let slot_duration = epoch_duration
+                            / epoch_sched.get_slots_in_epoch(interval_prices.end_epoch) as i32;
+                        now + slot_duration * (slots_diff * 8 / 10) as i32
+                    } else {
+                        // We can't estimate the epoch duration, use Solana's
+                        // `SLOT_MS`. Empirically, this is lower than than what
+                        // we observe in reality, so we'd cache for less time
+                        // than ideal.
+                        let ms_diff = slots_diff * SLOT_MS;
+                        now + chrono::Duration::milliseconds(ms_diff as i64)
+                    }
+                }
+            }
+        }
+    };
+
+    // Set minimum expiration time to 1 hour.
```

Comment:
1h is still quite a long time to have to wait for bad data to go away if we mess something up, let’s do something like 5 minutes? The same argument as for reducing it from 1y to 7d applies: those 11 additional requests are not what’s going to kill the app in cases of high load. It’s when we get into multiple requests per second territory that load _might_ become a problem (but actually I expect we can easily handle hundreds of requests/sec), and even caching for 1s in that case will reduce the load from an unbounded number of requests/sec to 1 request/sec. (Per set of query params, but if somebody is deliberately varying those, that’s a DoS attack.)

Reply:
Ok, 5 minutes seems reasonable, thanks!